File: src/main/java/dev/slne/discord/discord/interaction/modal/step/creator/report/steps/griefing/ReportTicketGriefingStep.java
package dev.slne.discord.discord.interaction.modal.step.creator.report.steps.griefing;

import dev.slne.discord.discord.interaction.modal.step.MessageQueue;
import dev.slne.discord.discord.interaction.modal.step.ModalComponentBuilder;
import dev.slne.discord.discord.interaction.modal.step.ModalSelectionStep;
import dev.slne.discord.message.RawMessages;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent;
import net.dv8tion.jda.api.interactions.components.selections.SelectOption;
import net.dv8tion.jda.api.interactions.components.text.TextInput;
import net.dv8tion.jda.api.interactions.components.text.TextInputStyle;

public class ReportTicketGriefingStep extends ModalSelectionStep {

  private static final String OPTION_SURVIVAL = "survival";
  private static final String OPTION_EVENT = "event";

  private static final String WORLD_INPUT = "world";
  private static final String XYZ_INPUT = "x y z";
  private static final String WHAT_GRIEFED_INPUT = "what-griefed";
  private static final String ADDITIONAL_INFORMATION_INPUT = "additional-information";

  private String world, xYZ, whatGriefed, additionalInformation;

  public ReportTicketGriefingStep() {
    super(RawMessages.get("modal.report.step.selection.griefing.title"),
        SelectOption.of(
                RawMessages.get("modal.report.step.selection.griefing.survival.label"),
                OPTION_SURVIVAL
            )
            .withDescription(
                RawMessages.get("modal.report.step.selection.griefing.survival.description")
            ),
        SelectOption.of(
                RawMessages.get("modal.report.step.selection.griefing.event.label"),
                OPTION_EVENT
            )
            .withDescription(
                RawMessages.get("modal.report.step.selection.griefing.event.description")
            )
    );
  }

  @Override
  protected void buildModalComponents(ModalComponentBuilder builder) {
    builder.addComponent(
        TextInput.create(
                WORLD_INPUT,
                RawMessages.get("modal.report.step.griefing.input.world.label"),
                TextInputStyle.SHORT
            )
            .setRequired(true)
            .setPlaceholder(RawMessages.get("modal.report.step.griefing.input.world.placeholder"))
            .build()
    );
    builder.addComponent(
        TextInput.create(
                XYZ_INPUT,
                RawMessages.get("modal.report.step.griefing.input.xyz.label"),
                TextInputStyle.SHORT
            )
            .setRequired(true)
            .setMinLength(5)
            .setPlaceholder(RawMessages.get("modal.report.step.griefing.input.xyz.placeholder"))
            .build()
    );
    builder.addComponent(
        TextInput.create(
                WHAT_GRIEFED_INPUT,
                RawMessages.get("modal.report.step.griefing.input.what-griefed.label"),
                TextInputStyle.PARAGRAPH
            )
            .setRequired(true)
            .setPlaceholder(RawMessages.get(
                "modal.report.step.griefing.input.what-griefed.placeholder"))
            .build()
    );
    builder.addComponent(
        TextInput.create(
                ADDITIONAL_INFORMATION_INPUT,
                RawMessages.get("modal.report.step.griefing.input.additional-info.label"),
                TextInputStyle.PARAGRAPH
            )
            .setRequired(false)
            .setPlaceholder(RawMessages.get(
                "modal.report.step.griefing.input.additional-info.placeholder"))
            .build()
    );
  }

  @Override
  protected void verifyModalInput(ModalInteractionEvent event)
      throws ModalStepInputVerificationException {
    world = getRequiredInput(event, WORLD_INPUT);
    xYZ = getRequiredInput(event, XYZ_INPUT);
    whatGriefed = getRequiredInput(event, WHAT_GRIEFED_INPUT);
    additionalInformation = getOptionalInput(event, ADDITIONAL_INFORMATION_INPUT);
  }

  @Override
  protected void buildOpenMessages(MessageQueue messages, TextChannel channel) {
    messages.addMessage(
        RawMessages.get("modal.report.step.griefing.messages.server", getSelectedServer()));
    messages.addMessage(
        RawMessages.get("modal.report.step.griefing.messages.location", xYZ, world));
    messages.addEmptyLine();
    messages.addMessage("# " + RawMessages.get(
        "modal.report.step.griefing.input.what-griefed.label"));
    messages.addMessage("> " + whatGriefed);

    if (additionalInformation != null && !additionalInformation.isBlank()) {
      messages.addEmptyLine();
      messages.addMessage("## " + RawMessages.get(
          "modal.report.step.griefing.input.additional-info.label"));
      messages.addMessage("> " + additionalInformation);
    }
  }

  private String getSelectedServer() {
    return switch (getSelected()) {
      case OPTION_SURVIVAL ->
          RawMessages.get("modal.report.step.selection.griefing.survival.label");
      case OPTION_EVENT -> RawMessages.get("modal.report.step.selection.griefing.event.label");
      default -> throw new IllegalStateException("Unexpected value: " + getSelected());
    };
  }
}

File: src/main/kotlin/dev/slne/discord/discord/interaction/modal/step/creator/report/steps/griefing/ReportTicketGriefingStep.kt
package dev.slne.discord.discord.interaction.modal.step.creator.report.steps.griefing

import dev.slne.discord.discord.interaction.modal.step.MessageQueue
import dev.slne.discord.discord.interaction.modal.step.ModalSelectionStep
import dev.slne.discord.message.translatable
import net.dv8tion.jda.api.components.label.Label
import net.dv8tion.jda.api.components.selections.SelectOption
import net.dv8tion.jda.api.components.textinput.TextInput
import net.dv8tion.jda.api.components.textinput.TextInputStyle
import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent
import net.dv8tion.jda.api.interactions.components.selections.StringSelectInteraction
import net.dv8tion.jda.api.modals.Modal

private const val OPTION_SURVIVAL_01 = "survival-1"
private const val OPTION_SURVIVAL_02 = "survival-2"
private const val OPTION_EVENT = "event"

private const val WORLD_INPUT = "world"
private const val XYZ_INPUT = "x y z"
private const val WHAT_GRIEFED_INPUT = "what-griefed"
private const val ADDITIONAL_INFORMATION_INPUT = "additional-information"

class ReportTicketGriefingStep : ModalSelectionStep(
    translatable("modal.report.step.selection.griefing.title"),
    SelectOption.of(
        translatable("modal.report.step.selection.griefing.survival1.label"),
        OPTION_SURVIVAL_01
    ).withDescription(translatable("modal.report.step.selection.griefing.survival.description")),
    SelectOption.of(
        translatable("modal.report.step.selection.griefing.survival2.label"),
        OPTION_SURVIVAL_02
    ).withDescription(translatable("modal.report.step.selection.griefing.survival.description")),
    SelectOption.of(
        translatable("modal.report.step.selection.griefing.event.label"),
        OPTION_EVENT
    ).withDescription(translatable("modal.report.step.selection.griefing.event.description"))
) {
    private var world: String? = null
    private var xYZ: String? = null
    private var whatGriefed: String? = null
    private var additionalInformation: String? = null

    override fun buildModalComponents(builder: Modal.Builder, interaction: StringSelectInteraction) {
        builder.addComponents(
            Label.of(
                translatable("modal.report.step.griefing.input.world.label"),
                TextInput.create(WORLD_INPUT, TextInputStyle.SHORT)
                    .setPlaceholder(translatable("modal.report.step.griefing.input.world.placeholder"))
                    .setRequired(true)
                    .build()
            ),
            Label.of(
                translatable("modal.report.step.griefing.input.xyz.label"),
                TextInput.create(XYZ_INPUT, TextInputStyle.SHORT)
                    .setPlaceholder(translatable("modal.report.step.griefing.input.xyz.placeholder"))
                    .setRequired(true)
                    .setMinLength(5)
                    .build()
            ),
            Label.of(
                translatable("modal.report.step.griefing.input.what-griefed.label"),
                TextInput.create(WHAT_GRIEFED_INPUT, TextInputStyle.PARAGRAPH)
                    .setPlaceholder(translatable("modal.report.step.griefing.input.what-griefed.placeholder"))
                    .setRequired(true)
                    .build()
            ),
            Label.of(
                translatable("modal.report.step.griefing.input.additional-info.label"),
                TextInput.create(ADDITIONAL_INFORMATION_INPUT, TextInputStyle.PARAGRAPH)
                    .setPlaceholder(translatable("modal.report.step.griefing.input.additional-info.placeholder"))
                    .setRequired(false)
                    .build()
            )
        )
    }

    override suspend fun verifyModalInput(event: ModalInteractionEvent) {
        world = event[WORLD_INPUT]
        xYZ = event[XYZ_INPUT]
        whatGriefed = event[WHAT_GRIEFED_INPUT]
        additionalInformation = getOptionalInput(event, ADDITIONAL_INFORMATION_INPUT)
    }

    override fun MessageQueue.buildOpenMessages(thread: ThreadChannel) {
        addMessage(
            translatable("modal.report.step.griefing.messages.server", selectedServer)
        )
        addMessage(
            translatable("modal.report.step.griefing.messages.location", xYZ, world)
        )
        addEmptyLine()
        addMessage(
            "# " + translatable(
                "modal.report.step.griefing.input.what-griefed.label"
            )
        )
        addMessage("> $whatGriefed")

        if (!additionalInformation.isNullOrEmpty()) {
            addEmptyLine()
            addMessage(
                "## " + translatable(
                    "modal.report.step.griefing.input.additional-info.label"
                )
            )
            addMessage("> $additionalInformation")
        }
    }

    private val selectedServer: String
        get() = when (selected) {
            OPTION_SURVIVAL_01 -> translatable("modal.report.step.selection.griefing.survival1.label")
            OPTION_SURVIVAL_02 -> translatable("modal.report.step.selection.griefing.survival2.label")
            OPTION_EVENT -> translatable("modal.report.step.selection.griefing.event.label")
            else -> throw IllegalStateException("Unexpected value: $selected")
        }
}
